src/index.ts

```diff
@@ -1,17 +1,14 @@
 import shallowEqualArrays, { validArrayValue } from "./arrays";
 import shallowEqualObjects, { validObjectValue } from "./objects";
 
-type Comparable<T> = Record<string, T> | T[] | null | undefined;
+type Comparable = Record<string, any> | any[] | null | undefined;
 
-function shallowEqual<T extends Comparable<T>>(a: T, b: T): boolean {
+function shallowEqual<T extends Comparable>(a: T, b: T): boolean {
   if (Array.isArray(a) || Array.isArray(b)) {
-    return shallowEqualArrays(a as validArrayValue<T>, b as validArrayValue<T>);
+    return shallowEqualArrays(a as validArrayValue, b as validArrayValue);
```

Comment:
Looking at this again, we're lying to ourselves here by using `as`.
If `a` is an array, and `b` is an object, we're telling TypeScript here that `b` is an array, which is wrong.
I think what we want here is TypeScript to complain if users pass `a` and an array and `b` as an object (or vice versa). This can be achieved with [function overloads](https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads) like [this](https://www.typescriptlang.org/play?#code/C4TwDgpgBAbghgGwJYBMCCAnDcQDVECu0AvFHAHYgDaAulAD5TkEIINQHkoQBmS5EFAG4AUKEixEqAPIAjAFYQAxsHwIiUUgCVlAewwoAPAGdgGfgHMANGUoA+ds1btO3PgOEiRPTiqS7yKGMAC0QEXQB3AFEARwJEAAo4AC5JZHQsHDUiG1lU+HTMbDxCCABKVNldXQQIClEfcj8AoNDWSNj4hCT8qRQ5RRVsiFze9IHlVVKKqCqauvIG32B-QJCwjrjElLTUIqzS9gKZBUnh0d2M4uGjvomh6agAbxEoKAB6d6gAFWkAEWkrygGAgwAIGECPEQxggogAvl5PlBdABrETrdrRLbdKgARhsACYaDYqABmGwAFhoZVEGPCWK6CSeO3xs1SBLhNmZqXJbKgFLhNMRXxCuhYKCgECw+lyBGAUCQ8pQuggxnIAHJ5bJlHACDCoHjCXQlBRZtA4MZjEgLB4oMBdJd7lN1BB0W16Z1EoaoESuTteXl+YLREiAHoAfi8IiAA), but we have another issue here: you can assign arrays to `Record<string, any>`:
```
const obj: Record<string, any> = [1, 2]; // no error!
```
I'm not sure at this point how to define a type that can't be assigned arrays. See [my question on StackOverflow](https://stackoverflow.com/q/75129453/247243).

Lastly, I believe that if we define types properly here, we shouldn't need the `as`.